In [220]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
from IPython.core.display import HTML, display

import time

In [227]:
pages_2016 = 154
pages_2017 = 147
pages_2018 = 85
pages_2019 = 113
pgs_half_2016 = 898
pgs_half_2017 = 892
pgs_half_2018 = 598
pgs_half_2019 = 690

pages = [pages_2016, pages_2017, pages_2018, pages_2019, 
         pgs_half_2016, pgs_half_2017, pgs_half_2018, pgs_half_2019]

urlbase_2019 = 'https://www.runrocknroll.com/Events/Nashville/The-Races/Marathon/2019-Results?page='
urlbase_2018 = 'https://www.runrocknroll.com/Events/Nashville/The-Races/Marathon/2018-Results?page='
urlbase_2017 = 'https://www.runrocknroll.com/Events/Nashville/The-Races/Marathon/2017-Results?page='
urlbase_2016 = 'https://www.runrocknroll.com/Events/Nashville/The-Races/Marathon/2016-Results?page='
urlbase_half_2019 = 'https://www.runrocknroll.com/Events/Nashville/The-Races/Half-Marathon/2019-Results?page='
urlbase_half_2018 = 'https://www.runrocknroll.com/Events/Nashville/The-Races/Half-Marathon/2018-Results?page='
urlbase_half_2017 = 'https://www.runrocknroll.com/Events/Nashville/The-Races/Half-Marathon/2017-Results?page='
urlbase_half_2016 = 'https://www.runrocknroll.com/Events/Nashville/The-Races/Half-Marathon/2016-Results?page='

urls = [urlbase_2016, urlbase_2017, urlbase_2018, urlbase_2019, 
         urlbase_half_2016, urlbase_half_2017, urlbase_half_2018, urlbase_half_2019]

full_2016 = pd.DataFrame()
full_2017 = pd.DataFrame()
full_2018 = pd.DataFrame()
full_2019 = pd.DataFrame()
half_2016 = pd.DataFrame()
half_2017 = pd.DataFrame()
half_2018 = pd.DataFrame()
half_2019 = pd.DataFrame()

df_list = [full_2016, full_2017, full_2018, full_2019, 
           half_2016, half_2017, half_2018, half_2019]

In [229]:
def get_result_page(url_base,page):
    
    url = url_base+str(page)
    
    result = requests.post(url)

    soup = BeautifulSoup(result.text,'lxml')

    tables = soup.find_all('table', attrs={'class': 'table table-responsive table-bordered'})
    
    result_list = pd.read_html(str(tables[0]))
    
    return result_list[0]


def get_race_results(year, category='Marathon'):
    
    df_race = pd.DataFrame()
    
#     if category=='Marathon':
        
#         index = int(year)-2016
        
#     elif category=='Half-Marathon':
    
#         index = int(year)-2016+4
    
    for i in range(pages[index]): # It will look all the pages
        
        page = i+1 # Page is starting from 1, ends with pages[index]

        df_race = pd.concat([df_race, get_result_page(urls[index],page)], ignore_index=True)
            
    return df_race

In [230]:
start_time = time.time()

for j in range(len(df_list)): # It will do the same for all race results for each marathon and year
    
    index = j

    url = urls[index]
    
    year = url.split('/')[7].split('-')[0]
    
    category = url.split('/')[6]
    
    df_list[index] = get_race_results(year,category)
    
    df_list[index]['year'] = year
    
    df_list[index]['category'] = category
    
    j = j+1

full_2016, full_2017, full_2018, full_2019, half_2016, half_2017, half_2018, half_2019 = df_list # Each df belongs to a race result. This will assign the values to each.

end_time = time.time()

print('It took '+ str((end_time-start_time)/60) + ' minutes to pull all results for these 8 races.')

# df_list

It took 20.2180999994 minutes to pull all results for these 8 races.


With get_result_page, get_race_results functions (keeping if elif for category and pd.append to combine tables):
It took 23.1160000006 minutes to pull all results for these 8 races. 

It decreased from 31 mins by defining these two functions. 

With get_result_page, get_race_results functions (shooting if elif and using pd.concat instead of pd.append to combine tables): It took 20.2180999994 minutes to pull all results for these 8 races.

It decreased from 23 mins by omitting if/elif and using pd.concat instead of .append.

In [126]:
full_2016.to_csv('data/full_2016.csv',index=False, encoding='utf-8')
full_2017.to_csv('data/full_2017.csv',index=False, encoding='utf-8')
full_2018.to_csv('data/full_2018.csv',index=False, encoding='utf-8')
full_2019.to_csv('data/full_2019.csv',index=False, encoding='utf-8')
half_2016.to_csv('data/half_2016.csv',index=False, encoding='utf-8')
half_2017.to_csv('data/half_2017.csv',index=False, encoding='utf-8')
half_2018.to_csv('data/half_2018.csv',index=False, encoding='utf-8')
half_2019.to_csv('data/half_2019.csv',index=False, encoding='utf-8')

In [218]:
full_2016.head()

Overall   Bib               Name      Time  year  category
0        1     1     Scott Wietecha  02:25:42  2016  Marathon
1        2     3      Brian Shelton  02:34:43  2016  Marathon
2        3  1047  Christopher Capps  02:38:43  2016  Marathon
3        4  1052       Jason Grimes  02:45:06  2016  Marathon
4        5  1046        David Adams  02:47:28  2016  Marathon

In [219]:
for j in range(len(df_list)):
    index = j
    url = urls[index]
    year = url.split('/')[7].split('-')[0]
    category = url.split('/')[6]
    print(year+' '+category)
    print(df_list[index].info())
    j=j+1

2016 Marathon
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3842 entries, 0 to 3841
Data columns (total 6 columns):
Overall     3842 non-null int64
Bib         3842 non-null int64
Name        3842 non-null object
Time        3842 non-null object
year        3842 non-null object
category    3842 non-null object
dtypes: int64(2), object(4)
memory usage: 180.2+ KB
None
2017 Marathon
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3651 entries, 0 to 3650
Data columns (total 6 columns):
Overall     3651 non-null int64
Bib         3651 non-null int64
Name        3651 non-null object
Time        3651 non-null object
year        3651 non-null object
category    3651 non-null object
dtypes: int64(2), object(4)
memory usage: 171.2+ KB
None
2018 Marathon
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2115 entries, 0 to 2114
Data columns (total 6 columns):
Overall     2115 non-null int64
Bib         2115 non-null int64
Name        2115 non-null object
Time        2115 non-null object
year